client/auth/login.tsx

```diff
@@ -0,0 +1,241 @@
+import queryString from 'query-string'
+import React, { useEffect, useRef, useState } from 'react'
+import { useTranslation } from 'react-i18next'
+import {
+  PASSWORD_MINLENGTH,
+  USERNAME_MAXLENGTH,
+  USERNAME_MINLENGTH,
+  USERNAME_PATTERN,
+} from '../../common/constants'
+import { useForm } from '../forms/form-hook'
+import SubmitOnEnter from '../forms/submit-on-enter'
+import { composeValidators, maxLength, minLength, regex, required } from '../forms/validators'
+import { RaisedButton } from '../material/button'
+import { push } from '../navigation/routing'
+import LoadingIndicator from '../progress/dots'
+import { useAppDispatch, useAppSelector } from '../redux-hooks'
+import { useStableCallback } from '../state-hooks'
+import { logIn } from './action-creators'
+import {
+  AuthBody,
+  AuthBottomAction,
+  AuthCheckBox,
+  AuthContent,
+  AuthContentContainer,
+  AuthPasswordTextField,
+  AuthTextField,
+  AuthTitle,
+  BottomActionButton,
+  FieldRow,
+  ForgotActionButton,
+  LoadingArea,
+  RowEdge,
+  Spacer,
+} from './auth-content'
+import { redirectIfLoggedIn } from './auth-utils'
+import { UserErrorDisplay } from './user-error-display'
+
+interface LoginModel {
+  username: string
+  password: string
+  rememberMe: boolean
+}
+
+export function Login() {
+  const { t, i18n } = useTranslation()
+  const dispatch = useAppDispatch()
+  const auth = useAppSelector(s => s.auth)
+
+  const [isLoading, setIsLoading] = useState(false)
+  const [lastError, setLastError] = useState<Error>()
+
+  useEffect(() => {
+    redirectIfLoggedIn({ auth })
+  }, [auth])
+
+  const abortControllerRef = useRef<AbortController>()
+
+  const onFormSubmit = useStableCallback((model: LoginModel) => {
+    setIsLoading(true)
+    setLastError(undefined)
+
+    abortControllerRef.current?.abort()
+    abortControllerRef.current = new AbortController()
+
+    dispatch(
+      logIn(
+        {
+          username: model.username,
+          password: model.password,
+          remember: model.rememberMe,
+          locale: i18n.language,
```

Comment:
No, like I said before, the thing we should be sending to the server is *the locale as reported by the browser*. This is *the language that was chosen based on the available languages*.

You need to keep track of when the user has set an explicit language preference in this client (via the dropdown), so you can send in the following priorities:

1) The explicit language preference they picked (if nothing explicit was chosen in the dropdown, this should be undefined. I don't consider opening the dropdown but not choosing anything to be an explicit selection)
2) The current locale as reported by the browser